### ISSUE TITLE
Inlining PostgreSQL CHAR(len)[] types produces VARCHAR[] instead

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#11441]
This code:

```java
System.out.println(ctx.render(DSL.val(new String[] {"a", "b"}, SQLDataType.CHAR(1).getArrayDataType())));
System.out.println(ctx.renderInlined(DSL.val(new String[] {"a", "b"}, SQLDataType.CHAR(1).getArrayDataType())));
```

Produces the following output:

```sql
?::char(1)[]
cast('{"a","b"}' as varchar[])
```

The inlined version is wrong, the bind value version is correct

**Comment by lukaseder** [comment]
Looking into this now. The problem is really PostgreSQL specific as well as inline specific. The relevant logic in `org.jooq.impl.DefaultBinding.DefaultArrayBinding::sqlInline0` is:

```java
            if (REQUIRES_ARRAY_CAST.contains(ctx.dialect())) {

                // [#8933] In some cases, we cannot derive the cast type from
                //         array type directly
                Class<?> arrayType =
                    dataType.getType() == Object[].class
                  ? deriveArrayTypeFromComponentType(value)
                  : dataType.getType();

                ctx.render().visit(cast(inline(PostgresUtils.toPGArrayString(value)), arrayType));
            }
```

So, we're casting as the Java `Class<?>` type, not the SQL `DataType<?>` type

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.6 (https://github.com/jOOQ/jOOQ/issues/13313), and 3.15.10 (https://github.com/jOOQ/jOOQ/issues/13314)

**Comment by lukaseder** [comment]
This is likely to fix a few other problems on our issue tracker, for which there haven't been any failing integration tests, yet. I'll close the duplicates once I find them.
